iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -76,30 +77,39 @@ public async Task RunSampleAsync()
 
         private void InitializeClient()
         {
-            // If the client reports Connected status, it is already in operational state.
-            if (s_connectionStatus != ConnectionStatus.Connected
-                && _deviceConnectionStrings.Any())
+            if (ShouldClientBeInitialized(s_connectionStatus))
             {
+                // Allow a single thread to dispose and initialize the client instance.
                 lock (_initLock)
                 {
-                    _logger.LogDebug($"Attempting to initialize the client instance, current status={s_connectionStatus}");
-
-                    // If the device client instance has been previously initialized, then dispose it.
-                    // The s_wasEverConnected variable is required to store if the client ever reported Connected status.
-                    if (s_wasEverConnected && s_connectionStatus == ConnectionStatus.Disconnected)
+                    if (ShouldClientBeInitialized(s_connectionStatus))
                     {
-                        s_deviceClient?.Dispose();
-                        s_wasEverConnected = false;
+                        _logger.LogDebug($"Attempting to initialize the client instance, current status={s_connectionStatus}");
+
+                        // If the device client instance has been previously initialized, then dispose it.
+                        if (s_deviceClient != null)
+                        {
+                            s_deviceClient.Dispose();
+                            s_deviceClient = null;
+                        }
                     }
 
-                    s_deviceClient = DeviceClient.CreateFromConnectionString(_deviceConnectionStrings.First(), _transportType);
+                    var options = new ClientOptions
+                    {
+                        SdkAssignsMessageId = Shared.SdkAssignsMessageId.WhenUnset,
+                    };
+
+                    s_deviceClient = DeviceClient.CreateFromConnectionString(_deviceConnectionStrings.First(), _transportType, options);
                     s_deviceClient.SetConnectionStatusChangesHandler(ConnectionStatusChangeHandler);
                     s_deviceClient.OperationTimeoutInMilliseconds = (uint)s_operationTimeout.TotalMilliseconds;
+
+                    _logger.LogDebug($"Initialized the client instance.");
                 }
 
                 try
                 {
-                    // Force connection now
+                    // Force connection now.
+                    // OpenAsync() is an idempotent call, it has the same effect if called once or multiple times on the same client.
                     s_deviceClient.OpenAsync().GetAwaiter().GetResult();
```

Comment:
Not part of this review, but I hate that we do this. We should consider reworking the sample so we don't call async code from a sync method.

Reply:
I agree. With the connection status change fix that went in recently, we might be able to get rid of a couple of these workarounds.
I'll go back to the reconnection sample after this PR is in.